src/expr/src/scalar/like_pattern.rs

```diff
@@ -73,3 +147,774 @@ pub fn build_regex(pattern: &str, flags: &str) -> Result<Regex, EvalError> {
+                }
+            }
+        }
+        // No string search needed, we just use a prefix match on rest.
+        if !rest.starts_with(&self.suffix) {
+            return false;
+        }
+        rest = &rest[self.suffix.len()..];
+        match &self.next {
+            None => rest.len() == 0,
+            Some(matcher) => matcher.is_match(rest),
+        }
+    }
+}
+
+/// Breaks a LIKE pattern into a chain of sub-patterns.
+fn build_string_matcher(pattern: &str, escape: char) -> Result<StringMatcher, EvalError> {
+    let mut head = StringMatcher {
+        consume: 0,
+        many: false,
+        suffix: String::from(""),
+        next: None,
+    };
+    let mut tail = &mut head;
+    let mut in_wildcard = true;
+    let mut in_escape = false;
+
+    for c in pattern.chars() {
+        match c {
+            c if !in_escape && c == escape => {
+                in_escape = true;
+                in_wildcard = false;
+            }
+            '_' if !in_escape => {
+                if in_wildcard {
+                    tail.consume += 1;
+                } else {
+                    // start a new partial match
+                    tail.next = Some(Box::new(StringMatcher {
+                        consume: 1,
+                        many: false,
+                        suffix: String::from(""),
+                        next: None,
+                    }));
+                    tail = tail.next.as_deref_mut().unwrap();
+                    in_wildcard = true;
+                }
+            }
+            '%' if !in_escape => {
+                if in_wildcard {
+                    tail.many = true;
+                } else {
+                    // start a new partial match
+                    tail.next = Some(Box::new(StringMatcher {
+                        consume: 0,
+                        many: true,
+                        suffix: String::from(""),
+                        next: None,
+                    }));
+                    tail = tail.next.as_deref_mut().unwrap();
+                    in_wildcard = true;
+                }
+            }
+            c => {
+                tail.suffix.push(c);
+                in_escape = false;
+                in_wildcard = false;
+            }
+        }
+    }
+    if in_escape {
+        return Err(EvalError::UnterminatedLikeEscapeSequence);
+    }
+    Ok(head)
+}
+
+// Unit Tests
+
+#[cfg(test)]
+mod test {
+    use super::*;
+    use ore::str::StrExt;
+
+    #[test]
+    fn test_like() {
+        struct Input<'a> {
+            haystack: &'a str,
+            matches: bool,
+        }
+        struct Pattern<'a> {
+            needle: &'a str,
+            case_insensitive: bool,
+            escape: char,
+            inputs: Vec<Input<'a>>,
+        }
+
+        let test_cases = vec![
+            Pattern {
+                needle: "",
+                case_insensitive: false,
+                escape: '\\',
+                inputs: vec![
```

Comment:
You can save a lot of lines if you define a function that generates the vector of inputs from an array.

Something like this (I might be missing type annotations):
```
let gen_inputs = |inputs| inputs.into_iter().map(|(haystack, matches)| Input { haystack, matches } |).collect::<Vec<_>>()
```
or as a separately defined function within this test module.

You would invoke the function like:
```
Pattern {
  ...
  inputs: gen_inputs([("", true), ("foo", false)])
}
```

Reply:
I tried this, but I was unable to get the lambda to return a vec of the correct type. Instead I used a lambda to construct Input, continued to use `vec![...]`. It does clean it up somewhat, but not as much as you were hoping for. Take a look and let me know if you have any ideas.

Reply:
I think all of these proposals are just fine (including the current state!), though I'll note I never mind the explicitness of spelling out each field, like you had originally. The meaning of a bool literal is always hard to intuit when it's not immediately adjacent to a label.